File: 133/solution.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    Node* cloneGraph(Node* node) {
        if(node == NULL)return NULL;
        unordered_map<Node*, Node*> visited;
        
        return dfs(node, visited);
    }
    Node* dfs(Node* node, unordered_map<Node*, Node*>& visited){
        if(visited.find(node) == visited.end()){
            Node* ans = new Node(node->val);
            visited[node] = ans;
            for(auto nei:node->neighbors)
                ans->neighbors.push_back(dfs(nei, visited));
            return ans;
        }else{
            return visited[node];
        }
    }
};

File: 133/solution.py
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution(object):
    def cloneGraph(self, node):
        """
        :type node: Node
        :rtype: Node
        """
        node_copy = self.dfs(node, dict())
        return node_copy
    
    def dfs(self, node, dic):
        if not node: return None
        if node in dic: return dic[node]
        node_cp = Node(node.val, [])
        dic[node] = node_cp
        for n in node.neighbors:
            neighbors_cp = self.dfs(n, dic)
            if neighbors_cp:
                node_cp.neighbors.append(neighbors_cp)
        return node_cp
        
